tortoisebeacon/tortoisebeacon.go

```diff
@@ -0,0 +1,559 @@
+		}
+	}
+
+	return beacon, nil
+}
+
+// GetBeacon waits until a Tortoise Beacon value is ready for a certain epoch and returns it as []byte.
+func (tb *TortoiseBeacon) GetBeacon(epochNumber types.EpochID) []byte {
+	if err := tb.Wait(epochNumber); err != nil {
+		return nil
+	}
+
+	v, err := tb.Get(epochNumber)
+	if err != nil {
+		return nil
+	}
+
+	return v.Bytes()
+}
+
+// Wait waits until beacon for this epoch becomes ready.
+func (tb *TortoiseBeacon) Wait(epochID types.EpochID) error {
+	tb.waitUntilStarted()
+
+	if tb.tortoiseBeaconDB != nil {
+		if _, ok := tb.tortoiseBeaconDB.GetTortoiseBeacon(epochID); ok {
+			return nil
+		}
+	}
+
+	tb.beaconsMu.RLock()
+	ch, ok := tb.beaconsReady[epochID]
+	tb.beaconsMu.RUnlock()
+
+	if !ok {
+		return ErrBeaconNotCalculated
+	}
+
+	<-ch
+
+	return nil
+}
+
+func (tb *TortoiseBeacon) waitUntilStarted() {
+	select {
+	case <-tb.CloseChannel():
+		return
+	case <-tb.started:
+		return
+	}
+}
+
+func (tb *TortoiseBeacon) setStarted() {
+	tb.startedOnce.Do(func() {
+		close(tb.started)
+	})
+}
+
+func (tb *TortoiseBeacon) cleanupLoop() {
+	ticker := time.NewTicker(cleanupInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-tb.CloseChannel():
+			return
+		case <-ticker.C:
+			tb.cleanup()
+		}
+	}
+}
+
+func (tb *TortoiseBeacon) cleanup() {
+	// TODO(nkryuchkov): implement a better solution, consider https://github.com/golang/go/issues/20135
+	tb.beaconsMu.Lock()
+	defer tb.beaconsMu.Unlock()
+
+	for e := range tb.beacons {
+		if tb.epochIsOutdated(e) {
+			delete(tb.beacons, e)
+			delete(tb.beaconsReady, e)
+		}
+	}
+}
+
+func (tb *TortoiseBeacon) epochIsOutdated(epoch types.EpochID) bool {
+	tb.layerMu.Lock()
+	lastEpoch := tb.lastLayer.GetEpoch()
+	tb.layerMu.Unlock()
+
+	return lastEpoch-epoch > cleanupEpochs
+}
+
+// listens to new layers.
+func (tb *TortoiseBeacon) listenLayers() {
+	for {
+		select {
+		case <-tb.CloseChannel():
+			tb.setStarted()
+
+			return
+
+		case layer := <-tb.layerTicker:
+			tb.Log.With().Info("Received tick",
+				log.Uint64("layer", uint64(layer)))
+
+			go tb.handleLayer(layer)
+		}
+	}
+}
+
+// the logic that happens when a new layer arrives.
+// this function triggers the start of new CPs.
+func (tb *TortoiseBeacon) handleLayer(layer types.LayerID) {
+	tb.layerMu.Lock()
+	if layer > tb.lastLayer {
+		tb.lastLayer = layer
+	}
+
+	tb.layerMu.Unlock()
+
+	epoch := layer.GetEpoch()
+	tb.Log.With().Info("tortoise beacon got tick",
+		log.Uint64("layer", uint64(layer)),
+		log.Uint64("epoch_id", uint64(epoch)))
+
+	tb.handleEpoch(epoch)
+}
+
+func (tb *TortoiseBeacon) handleEpoch(epoch types.EpochID) {
+	if epoch.IsGenesis() {
+		tb.Log.With().Info("not starting tortoise beacon since we are in genesis epoch",
+			log.Uint64("epoch_id", uint64(epoch)))
+
+		return
+	}
+
+	tb.Log.With().Info("Handling epoch",
+		log.Uint64("epoch_id", uint64(epoch)))
+
+	tb.beaconsMu.Lock()
+	if _, ok := tb.seenEpochs[epoch]; ok {
```

Comment:
what happens when a miner joins the network in the middle of the epoch? will it start sending proposals? as I recall this consensus protocol should start at the beginning of the epoch, if not - wait for the start of the next one